### PR TITLE
Mkulakow/kfs raw input contents

### PR DESCRIPTION
**Pull request by michalkulakowski**

### CODE REVIEW

File: src/binaryutils.cpp [hunk 1]
```diff
@@ -214,12 +215,13 @@ static Status validateTensor(const std::shared_ptr<TensorInfo>& tensorInfo,
...
     }
-
     return StatusCode::OK;
 }
 
 static Status validateTensor(const std::shared_ptr<TensorInfo>& tensorInfo,
-    const ::KFSRequest::InferInputTensor& src) {
+    const ::KFSRequest::InferInputTensor& src,
+    const std::string* buffer) {
+    SPDLOG_ERROR("2");
```
Comment (atobiszei) [inline src/binaryutils.cpp:224]: debug log

File: src/binaryutils.cpp [hunk 2]
```diff
@@ -232,24 +234,35 @@ static Status validateTensor(const std::shared_ptr<TensorInfo>& tensorInfo,
...
             tensorInfo->getBatchSize().has_value() ? tensorInfo->getBatchSize().value().toString() : std::string{"none"},
             src.contents().bytes_contents_size());
         return StatusCode::INVALID_BATCH_SIZE;
     }
 
-    for (int i = 0; i < src.contents().bytes_contents_size(); i++) {
-        if (src.contents().bytes_contents(i).size() <= 0) {
-            return StatusCode::BYTES_CONTENTS_EMPTY;
-        }
+    if (content_size <= 0) {
```
Comment (atobiszei) [inline src/binaryutils.cpp:246]: +spdlog with explanation

Comment (atobiszei) [reply src/binaryutils.cpp:246]: check if checkBatchSizeMismatch will always fail when input batchsize is zero. Tensor info can have 0 as a dynamic bs

File: src/binaryutils.cpp [hunk 3]
```diff
@@ -232,24 +234,35 @@ static Status validateTensor(const std::shared_ptr<TensorInfo>& tensorInfo,
...
+        return StatusCode::BYTES_CONTENTS_EMPTY;
     }
 
-    if (src.contents().bytes_contents_size() <= 0) {
-        return StatusCode::BYTES_CONTENTS_EMPTY;
+    if(buffer == nullptr) {
+        for (int i = 0; i < src.contents().bytes_contents_size(); i++) {
+            if (src.contents().bytes_contents(i).size() <= 0) {
+                return StatusCode::BYTES_CONTENTS_EMPTY;
```
Comment (atobiszei) [inline src/binaryutils.cpp:253]: +spdlog info that i-th image of batch have content size equal 0

File: src/binaryutils.cpp [hunk 4]
```diff
@@ -232,24 +234,35 @@ static Status validateTensor(const std::shared_ptr<TensorInfo>& tensorInfo,
...
-            return StatusCode::BYTES_CONTENTS_EMPTY;
-        }
+    if (content_size <= 0) {
+        return StatusCode::BYTES_CONTENTS_EMPTY;
     }
 
-    if (src.contents().bytes_contents_size() <= 0) {
-        return StatusCode::BYTES_CONTENTS_EMPTY;
+    if(buffer == nullptr) {
+        for (int i = 0; i < src.contents().bytes_contents_size(); i++) {
```
Comment (atobiszei) [inline src/binaryutils.cpp:251]:
```suggestion
        for (int i = 0; i < batchSize; i++) {
```

File: src/binaryutils.cpp [hunk 5]
```diff
@@ -340,16 +353,15 @@ static Status convertTensorToMatsMatchingTensorInfo(const TensorType& src, std::
     bool resizeSupported = isResizeSupported(tensorInfo);
     bool enforceResolutionAlignment = !resizeSupported;
 
-    for (int i = 0; i < getBinaryInputsSize(src); i++) {
-        cv::Mat image = convertStringToMat(getBinaryInput(src, i));
+    for (int i = 0; i < ((buffer == nullptr) ? getBinaryInputsSize(src) : 1); i++) {
```
Comment (atobiszei) [inline src/binaryutils.cpp:356]: move calculation of bs above

File: src/test/deserialization_tests.cpp [hunk 6]
```diff
@@ -430,7 +430,7 @@ TEST_P(DeserializeKFSTensorProtoNegative, ShouldReturnNullptrForPrecision) {
 TEST_P(DeserializeKFSTensorProto, ShouldReturnValidTensor) {
     auto [testedPrecision, getInputFromRawInputContents] = GetParam();
     std::string* bufferPtr = (getInputFromRawInputContents ? &buffer : nullptr);
-    if (!getInputFromRawInputContents && (ovms::Precision::FP16 == testedPrecision)) {
+    if ((!getInputFromRawInputContents && (ovms::Precision::FP16 == testedPrecision)) || (ovms::Precision::BIN == testedPrecision)) {
```
Comment (atobiszei) [inline src/test/deserialization_tests.cpp:433]: You move bin to supported yet here you skip the test.

File: src/binaryutils.cpp [hunk 7]
```diff
@@ -73,8 +73,8 @@ static bool isPrecisionEqual(int matPrecision, ovms::Precision tensorPrecision)
 
 static cv::Mat convertStringToMat(const std::string& image) {
     OVMS_PROFILE_FUNCTION();
-    std::vector<unsigned char> data(image.begin(), image.end());
-    cv::Mat dataMat(data, true);
```
Comment (dkalinowski) [inline src/binaryutils.cpp:77]: [Mat](https://docs.opencv.org/4.x/d3/d63/classcv_1_1Mat.html#a23b182c4ffd46abe38f460c1480ae887) (const std::vector< _Tp > &vec, bool copyData=false)

_copyData_:
Flag to specify whether the underlying data of the STL vector should be copied to (true) or shared with (false) the newly constructed matrix. When the data is copied, the allocated buffer is managed using Mat reference counting mechanism. While the data is shared, the reference counter is NULL, and you should not deallocate the data until the matrix is destructed.

Why we do not copy it anymore?

File: src/binaryutils.cpp [hunk 8]
```diff
@@ -232,22 +234,31 @@ static Status validateTensor(const std::shared_ptr<TensorInfo>& tensorInfo,
...
             tensorInfo->getBatchSize().has_value() ? tensorInfo->getBatchSize().value().toString() : std::string{"none"},
             src.contents().bytes_contents_size());
         return StatusCode::INVALID_BATCH_SIZE;
     }
 
-    for (int i = 0; i < src.contents().bytes_contents_size(); i++) {
-        if (src.contents().bytes_contents(i).size() <= 0) {
-            return StatusCode::BYTES_CONTENTS_EMPTY;
-        }
+    if (batchSize <= 0) {
```
Comment (dkalinowski) [inline src/binaryutils.cpp:246]: Shouldnt this if be above `checkBatchSizeMismatch`? 

`checkBatchSizeMismatch` checks if metadata would accept the actual static `batchSize` from request. Passing 0 would collide with `any` (0) in metadata.

File: src/test/binaryutils_test.cpp [hunk 9]
```diff
@@ -597,4 +596,41 @@ INSTANTIATE_TEST_SUITE_P(
...
+    uint8_t* ptr = static_cast<uint8_t*>(tensor.data());
+    EXPECT_EQ(std::equal(ptr, ptr + tensor.get_size(), rgb_expected_tensor), true);
+}
+
+TEST_F(BinaryUtilsTestKFSRawInputsContents, Negative_batchSizeBiggerThan1) {
+    ov::Tensor tensor;
+
+    std::shared_ptr<TensorInfo> tensorInfo = std::make_shared<TensorInfo>("", ovms::Precision::U8, ovms::Shape{2, 1, 1, 3}, Layout{"NHWC"});
+
+    ASSERT_EQ(convertBinaryRequestTensorToOVTensor(this->requestTensor, tensor, tensorInfo, &this->buffer), ovms::StatusCode::INVALID_BATCH_SIZE);
```
Comment (dkalinowski) [inline src/test/binaryutils_test.cpp:634]: Is this single test enough to cover all negative cases with last parameter set to non nullptr?

File: check_coverage.bat [hunk 10]
```diff
@@ -6,7 +6,7 @@
 
 #Rhel
 MIN_LINES_COV=74.8
-MIN_FUNCTION_COV=75.7
+MIN_FUNCTION_COV=75.6
```
Comment (dkalinowski) [inline check_coverage.bat:9]: Coverage decreased, should we add more tests? @michalkulakowski

File: src/test/deserialization_tests.cpp (hunk 6 above)
Comment (atobiszei) [reply src/test/deserialization_tests.cpp:433]: Why we even need to move BIN here to skip?

**Review by atobiszei — APPROVED** [review]

**Pull request merged**